### PR TITLE
Implement QLinearMul in mlas

### PR DESCRIPTION
**Pull request by zhanghuanrong**
Operator will be added later.

### CODE REVIEW

File: onnxruntime/core/mlas/lib/qlmul.cpp [hunk 1]
```diff
@@ -0,0 +1,345 @@
...
+        const auto va_hi_i16x8 = _mm_sub_epi16(MlasShiftRightInt16<DataType>(_mm_unpackhi_epi8(va_i8x16, va_i8x16), 8), VectorZeroPointA);
+
+        if (!IsScalarB) {
+            const auto vb_i8x16 = _mm_loadu_si128((const MLAS_INT32X4*)InputB);
+            InputB += 16;
+            vb_lo_i16x8 = _mm_sub_epi16(MlasShiftRightInt16<DataType>(_mm_unpacklo_epi8(vb_i8x16, vb_i8x16), 8), VectorZeroPointB);
+            vb_hi_i16x8 = _mm_sub_epi16(MlasShiftRightInt16<DataType>(_mm_unpackhi_epi8(vb_i8x16, vb_i8x16), 8), VectorZeroPointB);
+        }
+
+        auto ab_lo_lo = _mm_mullo_epi16(va_lo_i16x8, vb_lo_i16x8);
```
[the remark this answers is not in the document]
Comment (zhanghuanrong) [reply onnxruntime/core/mlas/lib/qlmul.cpp:185]: Good Suggestion. Working on it.

File: onnxruntime/core/mlas/lib/qlmul.cpp [hunk 2]
```diff
@@ -0,0 +1,345 @@
...
+        auto r_hi_hi = _mm_add_epi32(_mm_cvtps_epi32(_mm_mul_ps(_mm_cvtepi32_ps(_mm_unpackhi_epi16(ab_hi_lo, ab_hi_hi)), VectorScaleRatio)), VectorZeroPointC);
+        const auto vc_hi_i16x8 = _mm_packs_epi32(r_hi_lo, r_hi_hi);
+
+        auto vc = MlasPackS16_128<DataType>(vc_lo_i16x8, vc_hi_i16x8);
+        _mm_storeu_si128((MLAS_INT32X4*)OutputC, vc);
+        OutputC += 16;
+        N -= 16;
+    }
+
+    if (N > 0) {
```
[the remark this answers is not in the document]
Comment (zhanghuanrong) [reply onnxruntime/core/mlas/lib/qlmul.cpp:203]: The perf loss should be small in that pattern. Will use this pattern as more as possible.

File: onnxruntime/core/mlas/lib/qlmul.cpp [hunk 3]
```diff
@@ -0,0 +1,345 @@
...
+        const typename SUI::i8x8_t VectorB0 = SUI::vmov_n_i8(*InputB);
+        vb0_s16x8 = SUI::vreinterpretq_s16_i16(SUI::vsubl_i8(VectorB0, VectorZeroPointB));
+        vb1_s16x8 = vb0_s16x8;
+    }
+
+    while (N >= 16) {
+        const typename SUI::i8x16_t VectorA0 = SUI::vld1q_i8(InputA);
+        InputA += 16;
+        const int16x8_t va0_s16x8 = SUI::vreinterpretq_s16_i16(SUI::vsubl_i8(SUI::vget_low_i8(VectorA0), VectorZeroPointA));
+        const int16x8_t va1_s16x8 = SUI::vreinterpretq_s16_i16(SUI::vsubl_i8(SUI::vget_high_i8(VectorA0), VectorZeroPointA));
```
[the remark this answers is not in the document]
Comment (zhanghuanrong) [reply onnxruntime/core/mlas/lib/qlmul.cpp]: Fixed

File: onnxruntime/core/mlas/lib/qlmul.cpp [hunk 4]
```diff
@@ -0,0 +1,345 @@
...
+        if (!IsScalarB) {
+            MlasCopyTailBytes(TailData, (const uint8_t*)InputB, N);
+            const auto vb_i8x16 = _mm_loadu_si128((const MLAS_INT32X4*)TailData);
+            vb_lo_i16x8 = _mm_sub_epi16(MlasShiftRightInt16<DataType>(_mm_unpacklo_epi8(vb_i8x16, vb_i8x16), 8), VectorZeroPointB);
+            vb_hi_i16x8 = _mm_sub_epi16(MlasShiftRightInt16<DataType>(_mm_unpackhi_epi8(vb_i8x16, vb_i8x16), 8), VectorZeroPointB);
+        }
+
+        auto ab_lo_lo = _mm_mullo_epi16(va_lo_i16x8, vb_lo_i16x8);
+        auto ab_lo_hi = _mm_mulhi_epi16(va_lo_i16x8, vb_lo_i16x8);
+        auto r_lo_lo = _mm_add_epi32(_mm_cvtps_epi32(_mm_mul_ps(_mm_cvtepi32_ps(_mm_unpacklo_epi16(ab_lo_lo, ab_lo_hi)), VectorScaleRatio)), VectorZeroPointC);
```
[the remark this answers is not in the document]
Comment (zhanghuanrong) [reply onnxruntime/core/mlas/lib/qlmul.cpp]: That case exists. Change to do the float add first, then convert back to int.

**Review by tracysh — APPROVED** [review]

**Pull request merged**